11/Zaubeerer/generators.py

```diff
@@ -0,0 +1,55 @@
+"""
+Turn the following unix pipeline into Python code using generators
+
+$ for i in ../*/*py; do grep ^import $i|sed 's/import //g' ; done | sort | uniq -c | sort -nr
+   4 unittest
+   4 sys
+   3 re
+   3 csv
+   2 tweepy
+   2 random
+   2 os
+   2 json
+   2 itertools
+   1 time
+   1 datetime
+"""
+
+import glob, os
+import re
+
+import collections
+
+def gen_files(pat):
+    """Finds all files"""
+
+    yield from glob.glob(pat)
+
+def gen_lines(files):
+    """Finds all lines"""
+    
+    for file in files:
+        with open(file) as f:
+            yield from f.readlines()
+
+def gen_grep(lines, pattern):
+    """Analyze whether lines match a certain pattern"""
+    for line in lines:
+        if line.startswith(pattern):
+            yield line[len(pattern):-1]
+
+def gen_count(lines):
+    """Count number of occurrences of specific patterns"""
+
+    yield collections.Counter(line for line in lines)
```

Comment:
yield from?

Reply:
If I use `yield from` I get only the keys of the `Counter`.

I wrote the test function such, that `gen_count()` is expected to return a `Counter`.

Is that wrong?

Reply:
This yields the whole thing at once. If that passes tests then that's on us (I would need to check), but if the collection is super large I think you want to yield them one by one, e.g.

```
def gen_count(lines):
    return (f'{cnt[1]:>7} {cnt[0]}' for cnt in Counter(lines).most_common())
```

or

```
    cnt = Counter(lines)  # Counter works on iterables and is not memory heavy
    yield from cnt.most_common()  # most_common sorts them descending
```

(examples from community branch of this repo!)

Reply:
Yeah, I also just checked it and found this appealing. 

There are also one-line solutions using regexs, neat!